From dc89d77f57eed2064bc43d4b242a911232e44690 Mon Sep 17 00:00:00 2001
From: Hsia-Jun Li <randy.li@synaptics.com>
Date: Wed, 25 Jun 2025 11:29:11 +0800
Subject: [PATCH 24/24] v4l2videodec: always drop cur pool for new resolution

The Synaptics VPU trusted app would determind whether
it can continues on the current buffer configuration.

Upstream-Status: Pending
Signed-off-by: Hsia-Jun Li <randy.li@synaptics.com>
---
 .../sys/v4l2/gstv4l2videodec.c                | 46 ++-----------------
 1 file changed, 3 insertions(+), 43 deletions(-)

diff --git a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2videodec.c b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2videodec.c
index 3cae2c82fb..143e4ef41d 100644
--- a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2videodec.c
+++ b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2videodec.c
@@ -745,50 +745,9 @@ gst_v4l2_video_dec_wait_for_src_ch (GstV4l2VideoDec * self)
 static void
 gst_v4l2_video_dec_src_ch_handle (GstV4l2VideoDec * self)
 {
-  GstVideoInfo info;
-  GstVideoCodecState *output_state;
-
-  if (!gst_v4l2_object_acquire_format (self->v4l2capture, &info)) {
-    gst_v4l2_object_stop (self->v4l2capture);
-    return;
-  }
-
-  output_state = gst_video_decoder_get_output_state (GST_VIDEO_DECODER(self));
-  if (!output_state) {
-    gst_v4l2_object_stop (self->v4l2capture);
-    return;
-  }
-
-  if (GST_VIDEO_INFO_FORMAT(&output_state->info) ==
-      GST_VIDEO_INFO_FORMAT(&info) &&
-      output_state->info.width == info.width &&
-      output_state->info.height == info.height)
-  {
-    GstV4l2BufferPool *pool;
-
-    gst_video_codec_state_unref (output_state);
-    GST_INFO_OBJECT (self, "same resolution, will restart the decoder");
-
-    pool = GST_V4L2_BUFFER_POOL (gst_v4l2_object_get_buffer_pool
-        (self->v4l2capture));
-    if (pool) {
-      gst_v4l2_buffer_pool_resume(pool);
-      gst_object_unref (pool);
-    } else {
-      return;
-    }
-
-    if (gst_v4l2_decoder_cmd (self->v4l2output, V4L2_DEC_CMD_START, 0)) {
-      if (g_atomic_int_get (&self->completing))
-        if (gst_v4l2_decoder_cmd (self->v4l2output, V4L2_DEC_CMD_STOP, 0))
-          g_atomic_int_set (&self->completing, FALSE);
-
-      return;
-    }
-  }
-  /* TODO: we didn't handle less buffer or memory requirement case */
-  gst_video_codec_state_unref (output_state);
+  /* handle less buffer or memory requirement case */
   gst_v4l2_object_stop (self->v4l2capture);
+  self->wait_for_source_change = FALSE;
 }
 
 static void
@@ -939,6 +898,7 @@ beach:
   if (ret == GST_V4L2_FLOW_RESOLUTION_CHANGE) {
     GST_VIDEO_DECODER_STREAM_LOCK (decoder);
     self->draining = TRUE;
+    self->wait_for_source_change = TRUE;
     GST_VIDEO_DECODER_STREAM_UNLOCK (decoder);
     GST_INFO_OBJECT (decoder, "Received resolution change");
     return;
-- 
2.43.0

